### ISSUE TITLE
Rule proposal: no-useless-optional-chain

### ISSUE THREAD

**Issue by tyilo** [typescript-eslint/typescript-eslint#10824]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [x] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [x] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [x] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [x] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

Using the optional chaining operator on a value that is guaranteed by the type system to not be `undefined` or `null` is confusing.

Either a normal property access should be done or the type should be changed to reflect that it can actually be `undefined` or `null`.

### Fail Cases

```TypeScript
type Foo = { x: { y: number } };

function getY(foo: Foo | undefined): number | undefined {
  return foo?.x?.y;
}
```

### Pass Cases

```TypeScript
type Foo = { x: { y: number } };

function getY(foo: Foo | undefined): number | undefined {
  return foo?.x.y;
}

// Alternatively:
type Foo2 = { x?: { y: number } };

function getY2(foo: Foo2 | undefined): number | undefined {
  return foo?.x?.y;
}
```

**Comment by kirkwaiblinger** [comment]
This exists! ([playground](https://typescript-eslint.io/play/#ts=5.7.2&fileType=.tsx&code=PTAEDMEMEsBsCgDe9SgC4E8AOBTUAxAe0NAF5RFQAPALgtAzoDsBXAWwCMcAnUAX34BueCggsmAYzTRCTUAHMcaAJoAKcMTpESAH1DiAJjnDQmOAwEpm7Lrz2Hjp8xVGpuSltzkbCAfgB0VAEYwqh88OHwIKBYkADOcUiimLgExGT0tPSMoKycPPxCIqjg4lIycooq6pppuvpMRiZmltb5dg1NTgYuqG4eXhDEAVT%2BIaKRqNEAgrBoPEyQ0gBuOLCMydh42gBMGZRBdJQ5ebaFfMKipZLSsgpKyjs1hFrEe-aNji1WuTYFH10Wr0%2Bu40J5vMNAsFQvwIiIcFQsIRuGh6HwgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRQA7iKaRwYAL4hNQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)) 

See [no-unnecessary-condition](https://typescript-eslint.io/rules/no-unnecessary-condition/)

**Comment by tyilo** [comment]
> This exists 😀! ([playground](https://typescript-eslint.io/play/#ts=5.7.2&fileType=.tsx&code=PTAEDMEMEsBsCgDe9SgC4E8AOBTUAxAe0NAF5RFQAPALgtAzoDsBXAWwCMcAnUAX34BueCggsmAYzTRCTUAHMcaAJoAKcMTpESAH1DiAJjnDQmOAwEpm7Lrz2Hjp8xVGpuSltzkbCAfgB0VAEYwqh88OHwIKBYkADOcUiimLgExGT0tPSMoKycPPxCIqjg4lIycooq6pppuvpMRiZmltb5dg1NTgYuqG4eXhDEAVT%2BIaKRqNEAgrBoPEyQ0gBuOLCMydh42gBMGZRBdJQ5ebaFfMKipZLSsgpKyjs1hFrEe-aNji1WuTYFH10Wr0%2Bu40J5vMNAsFQvwIiIcFQsIRuGh6HwgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRQA7iKaRwYAL4hNQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false))
> 
> See [no-unnecessary-condition](https://typescript-eslint.io/rules/no-unnecessary-condition/)

Would be nice if the playground defaulted to enabling the recommended lints, so I could have easily checked this 🙁

**Comment by kirkwaiblinger** [comment]
Hmmm, I often fill in the `extends` with the preset config that I want ([like so](https://typescript-eslint.io/play/#ts=5.7.2&fileType=.tsx&code=PTAEDMEMEsBsCgDe9SgC4E8AOBTUAxAe0NAF5RFQAPALgtAzoDsBXAWwCMcAnUAX34BueCggsmAYzTRCTUAHMcaAJoAKcMTpESAH1DiAJjnDQmOAwEpm7Lrz2Hjp8xVGpuSltzkbCAfgB0VAEYwqh88OHwIKBYkADOcUiimLgExGT0tPSMoKycPPxCIqjg4lIycooq6pppuvpMRiZmltb5dg1NTgYuqG4eXhDEAVT%2BIaKRqNEAgrBoPEyQ0gBuOLCMydh42gBMGZRBdJQ5ebaFfMKipZLSsgpKyjs1hFrEe-aNji1WuTYFH10Wr0%2Bu40J5vMNAsFQvwIiIcFQsIRuGh6HwgA&eslintrc=N4KABGBECmAeAu0B2ATAzpAXGA2pADgDYCuA5gJZKYAC8AnvtGgMYBO5%2B8AtE4ZfAHo08ds271GXZgAtozANbQUkALogAvkA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)), but that's not a very clearly surfaced feature. In this case, the rule you were looking for is part of `strict-type-checked`, though not `recommended-type-checked`. Maybe the [page on preset configs](https://typescript-eslint.io/users/configs/) could/should contain playground links with the configs enabled?

It's not obvious to me that there's a unique default that it would make sense to put in the playground, since we have multiple different configs for different use cases. 

Also - I guess having the recommended/strict configs enabled would help discover most common/standard rules, but it wouldn't really help for the rules that aren't part of a preset config... Wondering if there's a better mechanism for rule discovery? 

Curious what your thoughts are.

**Comment by tyilo** [comment]
> Hmmm, I often fill in the `extends` with the preset config that I want ([like so](https://typescript-eslint.io/play/#ts=5.7.2&fileType=.tsx&code=PTAEDMEMEsBsCgDe9SgC4E8AOBTUAxAe0NAF5RFQAPALgtAzoDsBXAWwCMcAnUAX34BueCggsmAYzTRCTUAHMcaAJoAKcMTpESAH1DiAJjnDQmOAwEpm7Lrz2Hjp8xVGpuSltzkbCAfgB0VAEYwqh88OHwIKBYkADOcUiimLgExGT0tPSMoKycPPxCIqjg4lIycooq6pppuvpMRiZmltb5dg1NTgYuqG4eXhDEAVT%2BIaKRqNEAgrBoPEyQ0gBuOLCMydh42gBMGZRBdJQ5ebaFfMKipZLSsgpKyjs1hFrEe-aNji1WuTYFH10Wr0%2Bu40J5vMNAsFQvwIiIcFQsIRuGh6HwgA&eslintrc=N4KABGBECmAeAu0B2ATAzpAXGA2pADgDYCuA5gJZKYAC8AnvtGgMYBO5%2B8AtE4ZfAHo08ds271GXZgAtozANbQUkALogAvkA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)), but that's not a very clearly surfaced feature. In this case, the rule you were looking for is part of `strict-type-checked`, though not `recommended-type-checked`. Maybe the [page on preset configs](https://typescript-eslint.io/users/configs/) could/should contain playground links with the configs enabled?

Is there a reason that this rule is not in the recommended set?

> It's not obvious to me that there's a unique default that it would make sense to put in the playground, since we have multiple different configs for different use cases.
> 
> Also - I guess having the recommended/strict configs enabled would help discover most common/standard rules, but it wouldn't really help for the rules that aren't part of a preset config... Wondering if there's a better mechanism for rule discovery?
> 
> Curious what your thoughts are.

Maybe some UI for selecting the different provided sets. A list with a checkbox for each?

**Comment by bradzacher** [comment]
Because the recommended set is about guaranteed best practices that all codebases will always want.

Removing unnecessary conditions is not something all codebase want because many codebases either have lax types, don't have `strictNullChecks` turned on, or practice defensive programming and purposely have "unnecessary" checks.

It is a best practice to use the rule, for sure, but it's still somewhat opinionated. Hence it sits in the `strict` config.

**Comment by tyilo** [comment]
@kirkwaiblinger 

Another thing which makes the playground harder to use that it only has a `.eslintrc` file, but no `eslint.config.mjs` file, which https://typescript-eslint.io/getting-started/ and https://typescript-eslint.io/users/configs shows how to use.

How would I know to to take either
```javascript
export default tseslint.config(
  tseslint.configs.strictTypeChecked,
);
```

or

```javascript
module.exports = {
  extends: [
    'plugin:@typescript-eslint/strict-type-checked',
  ],
};
```

and convert it into the following?

```
{
  "extends": ["plugin:@typescript-eslint/strict-type-checked"]
}
```

Maybe we should open a new issue for making the playground use the `eslint.config.mjs` format.

**Comment by kirkwaiblinger** [comment]
> [@kirkwaiblinger](https://github.com/kirkwaiblinger)
> 
> Another thing which makes the playground harder to use that it only has a `.eslintrc` file, but no `eslint.config.mjs` file, which https://typescript-eslint.io/getting-started/ and https://typescript-eslint.io/users/configs shows how to use.
> 
> How would I know to to take either
> 
> export default tseslint.config(
>   tseslint.configs.strictTypeChecked,
> );
> or
> 
> module.exports = {
>   extends: [
>     'plugin:@typescript-eslint/strict-type-checked',
>   ],
> };
> and convert it into the following?
> 
> ```
> {
>   "extends": ["plugin:@typescript-eslint/strict-type-checked"]
> }
> ```
> 
> Maybe we should open a new issue for making the playground use the `eslint.config.mjs` format.

Agreed that the config file on the old eslintrc format from eslint<9 is sort of misleading, but unfortunately the fact that the eslint config is executable rather than static makes this a real problem - see https://github.com/typescript-eslint/typescript-eslint/issues/10239.  Again, open to ideas on how to make this better for users, though.

**Comment by kirkwaiblinger** [comment]
> Maybe some UI for selecting the different provided sets. A list with a checkbox for each?

This could be an idea... Maybe it makes sense to add some functionality similar to https://eslint.org/play, but reflect updates in the JSON format 🤔

**Comment by tyilo** [comment]
> Agreed that the config file on the old eslintrc format from eslint<9 is sort of misleading, but unfortunately the fact that the eslint config is executable rather than static makes this a real problem - see [#10239](https://github.com/typescript-eslint/typescript-eslint/issues/10239). Again, open to ideas on how to make this better for users, though.

I guess all the documentation pages should also include the `.eslintrc` format then...

**Comment by bradzacher** [comment]
You mean like our legacy eslint config page...?
https://typescript-eslint.io/getting-started/legacy-eslint-setup
